src/protocol/mod.rs

```diff
@@ -178,9 +200,19 @@ impl WriteState {
             Some(WriteSize { ref mut size, ref mut payload }) => {
                 match size.try_write(socket) {
                     Ok(WriterResult::Done) => {
-                        debug!("WriteSize {:?}: Transitioning to writing payload", token);
-                        let payload = mem::replace(payload, Writer::empty());
-                        Some(Some(WriteState::WriteData(payload)))
+                        let payload = mem::replace(payload, None);
+                        if let Some(payload) = payload {
+                            debug!("WriteSize {:?}: Transitioning to writing payload", token);
+                            Some(Some(WriteState::WriteData(payload)))
+                        } else {
+                            debug!("WriteSize {:?}: payload is empty. Done writing.", token);
+                            if outbound.is_empty() {
```

Comment:
Yeah I was thinking about doing something like that. What about the Err states though? It shouldn't insert readable() in those cases. The best thing to do is probably to refactor the `Option<Option<WriteState>>` to an enum that can cover that case.